File: LiftLog.Ui/Services/FeedIdentityService.cs
using System.Text;
using Google.Protobuf;
using LiftLog.Lib;
using LiftLog.Lib.Models;
using LiftLog.Lib.Services;
using LiftLog.Ui.Models;
using LiftLog.Ui.Store.Feed;
using Microsoft.Extensions.Logging;

namespace LiftLog.Ui.Services;

public class FeedIdentityService(
    IFeedApiService feedApiService,
    IEncryptionService encryptionService
)
{
    public async Task<ApiResult<FeedIdentity>> CreateFeedIdentityAsync(
        string? name,
        byte[]? profilePicture,
        bool publishBodyweight,
        bool publishPlan,
        bool publishWorkouts,
        ImmutableListValue<SessionBlueprint> currentPlan
    )
    {
        var response = await feedApiService.CreateUserAsync(new CreateUserRequest());
        if (!response.IsSuccess)
        {
            return ApiResult<FeedIdentity>.FromFailure(response);
        }

        var aesKey = await encryptionService.GenerateAesKeyAsync();
        var rsaKeyPair = await encryptionService.GenerateRsaKeysAsync();
        return await UpdateFeedIdentityAsync(
            id: response.Data.Id,
            lookup: response.Data.Lookup,
            password: response.Data.Password,
            aesKey: aesKey,
            rsaKeyPair: rsaKeyPair,
            name: name,
            profilePicture: profilePicture,
            publishBodyweight: publishBodyweight,
            publishPlan: publishPlan,
            publishWorkouts: publishWorkouts,
            currentPlan: currentPlan
        );
    }

    public async Task<ApiResult<FeedIdentity>> UpdateFeedIdentityAsync(
        Guid id,
        string lookup,
        string password,
        AesKey aesKey,
        RsaKeyPair rsaKeyPair,
        string? name,
        byte[]? profilePicture,
        bool publishBodyweight,
        bool publishPlan,
        bool publishWorkouts,
        ImmutableListValue<SessionBlueprint> currentPlan
    )
    {
        var privateKey = rsaKeyPair.PrivateKey;
        var (_, iv) = await encryptionService.SignRsa256PssAndEncryptAesCbcAsync(
            [1],
            aesKey,
            privateKey
        );
        byte[]? encryptedName = name is null
            ? null
            : (
                await encryptionService.SignRsa256PssAndEncryptAesCbcAsync(
                    Encoding.UTF8.GetBytes(name),
                    aesKey,
                    privateKey,
                    iv
                )
            ).EncryptedPayload;
        var encryptedProfilePicture = profilePicture is null
            ? null
            : (
                await encryptionService.SignRsa256PssAndEncryptAesCbcAsync(
                    profilePicture,
                    aesKey,
                    privateKey,
                    iv
                )
            ).EncryptedPayload;

        var currentProgram = (CurrentPlanDaoV1?)currentPlan;
        var encryptedPlan =
            currentProgram is null || !publishPlan
                ? null
                : (
                    await encryptionService.SignRsa256PssAndEncryptAesCbcAsync(
                        currentProgram.ToByteArray(),
                        aesKey,
                        privateKey,
                        iv
                    )
                ).EncryptedPayload;

        var result = await feedApiService.PutUserDataAsync(
            new PutUserDataRequest(
                Id: id,
                Password: password,
                EncryptedCurrentPlan: encryptedPlan,
                EncryptedName: encryptedName,
                EncryptedProfilePicture: encryptedProfilePicture,
                EncryptionIV: iv.Value,
                RsaPublicKey: rsaKeyPair.PublicKey.SpkiPublicKeyBytes
            )
        );
        if (!result.IsSuccess)
        {
            return ApiResult<FeedIdentity>.FromFailure(result);
        }

        return ApiResult.Success(
            new FeedIdentity(
                Id: id,
                Lookup: lookup,
                AesKey: aesKey,
                RsaKeyPair: rsaKeyPair,
                Password: password,
                Name: name,
                ProfilePicture: profilePicture,
                PublishBodyweight: publishBodyweight,
                PublishPlan: publishPlan,
                PublishWorkouts: publishWorkouts
            )
        );
    }

    public async Task<ApiResult> DeleteFeedIdentityAsync(FeedIdentity identity)
    {
        var response = await feedApiService.DeleteUserAsync(
            new DeleteUserRequest(Id: identity.Id, Password: identity.Password)
        );
        return response;
    }
}

File: app/services/feed-identity-service.ts
import { FeedApiService } from './feed-api';
import { EncryptionService } from './encryption-service';
import { AesKey, RsaKeyPair } from '@/models/encryption-models';
import { DeleteUserRequest } from '@/models/feed-api-models';
import { FeedIdentity } from '@/models/feed-models';
import { SessionBlueprint } from '@/models/blueprint-models';
import { toCurrentPlanDao } from '@/models/storage/conversions.to-dao';
import { LiftLog } from '@/gen/proto';
import { ApiResult } from '@/services/api-error';

export class FeedIdentityService {
  constructor(
    private feedApiService: FeedApiService,
    private encryptionService: EncryptionService,
  ) {}

  async createFeedIdentityAsync(
    name: string | undefined,
    profilePicture: Uint8Array | undefined,
    publishBodyweight: boolean,
    publishPlan: boolean,
    publishWorkouts: boolean,
    currentPlan: SessionBlueprint[],
  ): Promise<ApiResult<FeedIdentity>> {
    const response = await this.feedApiService.createUserAsync();
    if (!response.isSuccess()) {
      return ApiResult.fromFailure(response);
    }
    const aesKey = await this.encryptionService.generateAesKey();
    const rsaKeyPair = await this.encryptionService.generateRsaKeys();
    return this.updateFeedIdentityAsync(
      response.data.id,
      response.data.lookup,
      response.data.password,
      aesKey,
      rsaKeyPair,
      name,
      profilePicture,
      publishBodyweight,
      publishPlan,
      publishWorkouts,
      currentPlan,
    );
  }

  async updateFeedIdentityAsync(
    id: string,
    lookup: string,
    password: string,
    aesKey: AesKey,
    rsaKeyPair: RsaKeyPair,
    name: string | undefined,
    profilePicture: Uint8Array | undefined,
    publishBodyweight: boolean,
    publishPlan: boolean,
    publishWorkouts: boolean,
    currentPlan: SessionBlueprint[],
  ): Promise<ApiResult<FeedIdentity>> {
    const privateKey = rsaKeyPair.privateKey;
    const { iv } =
      await this.encryptionService.signRsa256PssAndEncryptAesCbcAsync(
        new Uint8Array([1]),
        aesKey,
        privateKey,
      );

    const encryptedName = name
      ? (
          await this.encryptionService.signRsa256PssAndEncryptAesCbcAsync(
            new TextEncoder().encode(name),
            aesKey,
            privateKey,
            iv,
          )
        ).encryptedPayload
      : undefined;

    const encryptedProfilePicture = profilePicture
      ? (
          await this.encryptionService.signRsa256PssAndEncryptAesCbcAsync(
            profilePicture,
            aesKey,
            privateKey,
            iv,
          )
        ).encryptedPayload
      : undefined;
    const currentPlanDao = toCurrentPlanDao(currentPlan);
    const currentPlanBytes =
      LiftLog.Ui.Models.CurrentPlanDaoV1.encode(currentPlanDao).finish();
    const encryptedCurrentPlan = publishPlan
      ? (
          await this.encryptionService.signRsa256PssAndEncryptAesCbcAsync(
            currentPlanBytes,
            aesKey,
            privateKey,
            iv,
          )
        ).encryptedPayload
      : undefined;

    const result = await this.feedApiService.putUserDataAsync({
      id,
      password,
      encryptedCurrentPlan,
      encryptedName,
      encryptedProfilePicture,
      encryptionIV: iv.value,
      rsaPublicKey: rsaKeyPair.publicKey.spkiPublicKeyBytes,
    });

    if (!result.isSuccess()) {
      return ApiResult.fromFailure(result);
    }

    return ApiResult.success(
      new FeedIdentity(
        id,
        lookup,
        aesKey,
        rsaKeyPair,
        password,
        name,
        profilePicture,
        publishBodyweight,
        publishPlan,
        publishWorkouts,
      ),
    );
  }

  async deleteFeedIdentityAsync(
    identity: FeedIdentity,
  ): Promise<ApiResult<void>> {
    const response = await this.feedApiService.deleteUserAsync({
      id: identity.id,
      password: identity.password,
    } as DeleteUserRequest);

    return response;
  }
}
